src/backend/distributed/master/citus_create_restore_point.c

```diff
@@ -0,0 +1,199 @@
+ * than the coordinator.
+ */
+Datum
+citus_create_restore_point(PG_FUNCTION_ARGS)
+{
+	text *restoreNameText = PG_GETARG_TEXT_P(0);
+	char *restoreNameString = NULL;
+	XLogRecPtr localRestorePoint = InvalidXLogRecPtr;
+	List *connectionList = NIL;
+
+	CheckCitusVersion(ERROR);
+	EnsureSuperUser();
+	EnsureCoordinator();
+
+	if (RecoveryInProgress())
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 (errmsg("recovery is in progress"),
+				  errhint("WAL control functions cannot be executed during recovery."))));
+	}
+
+	if (!XLogIsNeeded())
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("WAL level not sufficient for creating a restore point"),
+				 errhint(
+					 "wal_level must be set to \"replica\" or \"logical\" at server start.")));
+	}
+
+	restoreNameString = text_to_cstring(restoreNameText);
+	if (strlen(restoreNameString) >= MAXFNAMELEN)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("value too long for restore point (maximum %d characters)",
+						MAXFNAMELEN - 1)));
+	}
+
+	/* establish connections to all nodes before taking any locks */
+	connectionList = OpenConnectionsToAllNodes();
+
+	/*
+	 * Send a BEGIN to bust through pgbouncer. We won't actually commit since
+	 * that takes time. Instead we just close the connections and roll back,
+	 * which doesn't undo pg_create_restore_point.
+	 */
+	RemoteTransactionListBegin(connectionList);
+
+	/* DANGER: finish as quickly as possible after this */
+	BlockAllDistributedWrites();
+
+	/* do local restore point first to bail out early if something goes wrong */
+	localRestorePoint = XLogRestorePoint(restoreNameString);
+
+	/* run pg_create_restore_point on all nodes */
+	CreateRemoteRestorePoints(restoreNameString, connectionList);
+
+	PG_RETURN_LSN(localRestorePoint);
+}
+
+
+/*
+ * OpenConnectionsToAllNodes opens connections to all nodes and returns the list
+ * of connections.
+ */
+static List *
+OpenConnectionsToAllNodes(void)
+{
+	List *connectionList = NIL;
+	List *workerNodeList = NIL;
+	ListCell *workerNodeCell = NULL;
+
+	workerNodeList = ActiveWorkerNodeList();
+
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		int connectionFlags = FORCE_NEW_CONNECTION;
+		MultiConnection *connection = NULL;
+
+		connection = StartNodeConnection(connectionFlags, workerNode->workerName,
+										 workerNode->workerPort);
+		MarkRemoteTransactionCritical(connection);
+
+		connectionList = lappend(connectionList, connection);
+	}
+
+	FinishConnectionListEstablishment(connectionList);
+
+	return connectionList;
+}
+
+
+/*
+ * BlockAllDistributedWrites blocks all modifications to distributed tables
+ * by taking ShareRowExclusive locks on all distributed tables.
+ */
+static void
+BlockAllDistributedWrites(void)
+{
+	ListCell *distributedTableCell = NULL;
+	List *distributedTableList = DistributedTableList();
```

Comment:
It doesn't really help unless all transactions always modify tables in the same order.